### PR TITLE
Rewrote Java multi-language pipeline quickstart

### PR DESCRIPTION
**Pull request by pcoet**
**Please** add a meaningful description for your change here

Refreshed the content for the Java multi-language pipeline quickstart. See: http://apache-beam-website-pull-requests.storage.googleapis.com/22263/documentation/sdks/java-multi-language-pipelines/index.html

@chamikaramj
@ihji 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by ihji** [comment]
Probably worth mentioning that we also have higher-level [`DataframeTransform`](https://github.com/apache/beam/blob/master/sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/transforms/DataframeTransformTest.java#L51) so users don't need to manually specify Python DataframeTransform FQN.

**Comment by pcoet** [comment]
> Probably worth mentioning that we also have higher-level [`DataframeTransform`](https://github.com/apache/beam/blob/master/sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/transforms/DataframeTransformTest.java#L51) so users don't need to manually specify Python DataframeTransform FQN.

Thanks, Heejong. Done.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md [hunk 1]
```diff
@@ -19,162 +19,185 @@ limitations under the License.
...
-        --expansionServiceURL=localhost:9097 \
-        --experiments=beam_fn_api"```
+```
+
+`PythonExternalTransform` is a wrapper for invoking external Python functions.
+The
+[`from`](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/extensions/python/PythonExternalTransform.html#from-java.lang.String-java.lang.String-)
+method accepts two strings: 1) the fully qualified transform name; 2) an
+optional address and port number for the expansion service. The method returns
+the transform.
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:100]: returns a stub transforms for the Python cross-language transform that can be directly used in a Java pipeline.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md [hunk 2]
```diff
@@ -19,162 +19,185 @@ limitations under the License.
...
+[`withKwarg`](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/extensions/python/PythonExternalTransform.html#withKwarg-java.lang.String-java.lang.Object-)
+specifies a keyword argument for instantiating the Python cross-language
+transform. In this case, `withKwarg` is invoked twice, to specify a `func`
+argument and an `include_indexes` argument, and these arguments are passed to
+`DataframeTransform`.
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:105]: Add,
Note that the PythonExternalTransform API offers additional mechanism for speficying args and kwargs for Python cross-language transforms.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md [hunk 3]
```diff
@@ -19,162 +19,185 @@ limitations under the License.
 
 # Java multi-language pipelines quickstart
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:20]: I think we should call this "Java multi-language quickstart for using Python transforms" and change the title/URL etc. accordingly.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md [hunk 4]
```diff
@@ -19,162 +19,185 @@ limitations under the License.
...
+[Embedding DataFrames in a pipeline](/documentation/dsls/dataframes/overview/#embedding-dataframes-in-a-pipeline).
+
+## Run the Java pipeline
+
+If you want to customize the environment or use transforms not available in the
+default Beam SDK, you might need to run your own expansion service. In such
+cases, [start the expansion service](#advanced-start-an-expansion-service)
+before running your pipeline.
+
+For the general use case, you can simply run your multi-language pipeline using
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:128]: For running this example, you can ...

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md [hunk 5]
```diff
@@ -19,162 +19,185 @@ limitations under the License.
...
+--runner=DataflowRunner \
+--output=gs://${OUTPUT_BUCKET}/count \
+--region=${GCP_REGION} \
+--experiments=use_runner_v2"
+```
+
+The pipeline outputs a file with the results to
+**gs://$OUTPUT_BUCKET/count-00000-of-00001**.
+
+Note: For Beam 2.40.0, you also need to set `sdkHarnessContainerImageOverrides`.
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:153]: I don't think we should necessarily document this. If we do we should provide an example (how to push a container and use that).

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md [hunk 6]
```diff
@@ -19,162 +19,185 @@ limitations under the License.
...
+   ```
+   python apache_beam/runners/portability/expansion_service_main.py -p 18089 --fully_qualified_name_glob "*"
+   ```
+
+The command runs
+[expansion_service_main.py](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/portability/expansion_service_main.py), which starts the standard expansion service. You should see output similar to
+the following:
+
+```
+INFO:root:Default Python SDK image for environment is <IMAGE>
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:186]: These logs may change in the future and may not be consistent with what we document here.

**Review by chamikaramj — COMMENTED** [review]
Thanks David.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md [hunk 7]
```diff
@@ -19,162 +19,185 @@ limitations under the License.
...
+To apply a cross-language transform, your pipeline must specify it. Python
+transforms are identified by their fully qualified name. For example,
+`DataframeTransform` can be found in the `apache_beam.dataframe.transforms`
+package, so its fully qualified name is
+`apache_beam.dataframe.transforms.DataframeTransform`.
+The example pipeline,
+[PythonDataframeWordCount](https://github.com/apache/beam/blob/master/examples/multi-language/src/main/java/org/apache/beam/examples/multilanguage/PythonDataframeWordCount.java),
+passes this fully qualified name to
+[PythonExternalTransform](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/extensions/python/PythonExternalTransform.html).
+There's also a higher-level cross-language [DataframeTransform](https://github.com/apache/beam/blob/master/sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/transforms/DataframeTransform.java)
```
[the remark this answers is not in the document]
Comment (pcoet) [reply website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:69]: That's great, thanks. Done.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md [hunk 8]
```diff
@@ -19,162 +19,185 @@ limitations under the License.
...
+for Java, so you can use that instead of specifying the fully qualified name for
+the Python `DataframeTransform`.
+
+Here's the complete pipeline definition from the example:
```
[the remark this answers is not in the document]
Comment (pcoet) [reply website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:73]: Done.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md [hunk 9]
```diff
@@ -19,162 +19,185 @@ limitations under the License.
...
-    -Pportable-runner \
-    -Dexec.args=" \
-        --runner=PortableRunner \
-        --jobEndpoint=localhost:8099 \
-        --useExternal=true \
-        --expansionServiceURL=localhost:9097 \
-        --experiments=beam_fn_api"```
+```
+
+`PythonExternalTransform` is a wrapper for invoking external Python functions.
```
[the remark this answers is not in the document]
Comment (pcoet) [reply website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:95]: Done.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md (hunk 1 above)
Comment (pcoet) [reply website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:100]: Done.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md (hunk 3 above)
Comment (pcoet) [reply website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:20]: I would push back on this a little bit. The intro makes it clear what we're doing in this quickstart: "In the examples below, the multi-language pipeline is built with the Beam Java SDK, and the cross-language transform is built with the Beam Python SDK." It's analogous to what we're doing with the Python multi-lang quickstart. At some point in the future, we _might_ need to change both sets of title/URL and narrow the focus, but we're not there yet. I'd rather maintain the symmetry. But let me know...

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md (hunk 5 above)
Comment (pcoet) [reply website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:153]: Removed.

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md (hunk 6 above)
Comment (pcoet) [reply website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:186]: ACK. I removed the log output.

**Comment by pcoet** [comment]
Thanks, Cham! I made all the changes accept for the title/URL. I think we should leave that as is (see my comment). But let me know...

File: website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md (hunk 3 above)
Comment (chamikaramj) [reply website/www/site/content/en/documentation/sdks/java-multi-language-pipelines.md:20]: Sounds good. Thanks.

**Review by chamikaramj — APPROVED** [review]
LGTM.

**Pull request merged**